### PR TITLE
add MonocleMock

### PR DESCRIPTION
**Pull request by friedm**
Please review: @sah

Mock extension that wraps results in a Callback so we can yield on them. Makes testing nested o-routines easier. Props to original author @miketwo.

### CODE REVIEW

File: monocle/monocle_mock.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        cb(result)
+        return cb
+
+
+class MonocleMock(_MonocleCallableMixin, Mock):
+    ''' Use this in the same way you would a regular Python Mock, but for
+    monocle'd methods. For example:
+
+    THE OLD WAY:
+    @twistedtest_o
```
Comment (miketwo) [inline monocle/monocle_mock.py]: I think this decorator is sauce-specific.  Might need a more generic example. Any of the ones from the unit tests in this PR would work.

File: monocle/monocle_mock.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+from mock import CallableMixin, Mock, MagicMock
+
+from monocle.callback import Callback
+
+
+class _MonocleCallableMixin(CallableMixin):
+    def _mock_call(self, *args, **kwargs):
+        ''' Mock seems to run all returns through this method, so we intercept
+        the return value and wrap it in monocle magic. '''
+        result = super(_MonocleCallableMixin, self)._mock_call(*args, **kwargs)
```
Comment (miketwo) [inline monocle/monocle_mock.py]: One other thing: we might need to use Tox or something to check which versions of Python and `mock` this is valid for. It's messing around with some deep mock internals, and there's no API garuntee at that level.

Comment (friedm) [inline monocle/monocle_mock.py]: Travis seems to have tested against all supported python versions already https://github.com/saucelabs/monocle/blob/master/tox.ini#L3-L5 . I think we're good as long as the tests I added are adequate.

File: setup.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
 from monocle import VERSION
 
-install_requires = []
+install_requires = ['mock']
```
Comment (sah) [inline setup.py]: This is kind of unfortunate. Means we'll have to install mock in prod in order to use monocle there. Two possible solutions:

1) MonocleMock becomes its own package. Continues to require mock, but we don't have to install it in prod.

2) Leave this requirement out. MonocleMock will silently need mock to be installed to work, but we won't insist on mock being installed to use the rest of monocle. This is not the normal way to do things in python, but note that monocle already takes this approach to deal with its unusual dependency on either twisted or tornado.

File: monocle/monocle_mock.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        # Special case for when a person tries to print a MonocleMock directly.
+        # In this case the result is a string, and the content is something
+        # like <MonocleMock id='4470907728'>. So we look for that particular
+        # string and do NOT wrap it in a Monocle callback.
+        if isinstance(result, str) and "MonocleMock " in result:
```
Comment (sah) [inline monocle/monocle_mock.py]: Maybe the '__repr__' method appears somewhere in the args, and could provide a more foolproof way to check for this case? If not, I'd say at least tighten this to `.startswith('<MonocleMock ')`

File: setup.py (hunk 3 above)
Comment (friedm) [inline setup.py]: Ah, good point. I felt kind of bad adding the first explicit dependency :p
2 sounds good to me, I'll add a test to make sure MonocleMock fails somewhat helpfully if mock is missing.

File: monocle/monocle_mock.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
+try:
+    from mock import CallableMixin, Mock, MagicMock
+except ImportError:
+    raise ImportError('monocle_mock requires mock as an additional dependency')
```
Comment (friedm) [inline monocle/monocle_mock.py]: Tried this out manually and it seemed ok.

File: monocle/monocle_mock.py (hunk 4 above)
Comment (friedm) [inline monocle/monocle_mock.py]: The args here are just what was passed into the mocked call, so we don't have that information available AFAICT. Went the suggested `.startswith(...)` route.

**Pull request merged**